go/vt/vtctl/reparentutil/emergency_reparenter.go

```diff
@@ -255,150 +586,163 @@ func (erp *EmergencyReparenter) promoteNewPrimary(
+func (erp *EmergencyReparenter) identifyPrimaryCandidate(
+	intermediateSource *topodatapb.Tablet,
+	prevPrimary *topodatapb.Tablet,
+	validCandidates []*topodatapb.Tablet,
+	tabletMap map[string]*topo.TabletInfo,
+	opts EmergencyReparentOptions,
+) (candidate *topodatapb.Tablet, err error) {
+	defer func() {
+		if candidate != nil {
+			erp.logger.Infof("found better candidate - %v", candidate.Alias)
+		}
+	}()
 
-	tabletMap, err := erp.ts.GetTabletMapForShard(ctx, keyspace, shard)
-	if err != nil {
-		return vterrors.Wrapf(err, "failed to get tablet map for %v/%v: %v", keyspace, shard, err)
+	if opts.NewPrimaryAlias != nil {
+		// explicit request to promote a specific tablet
+		requestedPrimaryAlias := topoproto.TabletAliasString(opts.NewPrimaryAlias)
+		requestedPrimaryInfo, isFound := tabletMap[requestedPrimaryAlias]
+		if !isFound {
+			return nil, vterrors.Errorf(vtrpc.Code_INTERNAL, "candidate %v not found in the tablet map; this an impossible situation", requestedPrimaryAlias)
+		}
+		for _, validCandidate := range validCandidates {
+			if topoproto.TabletAliasEqual(validCandidate.Alias, opts.NewPrimaryAlias) {
+				return requestedPrimaryInfo.Tablet, nil
+			}
+		}
+		return nil, vterrors.Errorf(vtrpc.Code_ABORTED, "requested candidate %v is not in valid candidates list", requestedPrimaryAlias)
 	}
-
-	statusMap, primaryStatusMap, err := StopReplicationAndBuildStatusMaps(ctx, erp.tmc, ev, tabletMap, opts.WaitReplicasTimeout, opts.IgnoreReplicas, erp.logger)
-	if err != nil {
-		return vterrors.Wrapf(err, "failed to stop replication and build status maps: %v", err)
+	var (
+		preferredCandidates []*topodatapb.Tablet
+		neutralReplicas     []*topodatapb.Tablet
+	)
+	for _, candidate := range validCandidates {
+		promotionRule := PromotionRule(candidate)
+		if promotionRule == promotionrule.Must || promotionRule == promotionrule.Prefer {
+			preferredCandidates = append(preferredCandidates, candidate)
+		}
+		if promotionRule == promotionrule.Neutral {
+			neutralReplicas = append(neutralReplicas, candidate)
+		}
 	}
 
-	if err := topo.CheckShardLocked(ctx, keyspace, shard); err != nil {
-		return vterrors.Wrapf(err, "lost topology lock, aborting: %v", err)
+	// So we already have an intermediate source. What if our intermediate source was a rdonly?
+	// So we will try to improve our candidate selection.
+	// Are there any replicas with better promotion rules?
+	// Maybe we actually promoted such a replica. Does that mean we should keep it?
+	// Maybe we promoted a "neutral", and some "prefer" server is available.
+	// Maybe we promoted a "prefer_not"
+	// Maybe we promoted a server in a different cell than the primary
+	// There's many options. We may wish to replace the server we promoted with a better one.
+
+	// check whether the one we promoted is in the same cell and belongs to the preferred candidates list
+	candidate = findPossibleCandidateFromListWithRestrictions(intermediateSource, prevPrimary, preferredCandidates, true, true)
```

Comment:
Similar to the other bools comments, we should annotate so it makes it easier to read.

Reply:
At the risk of repeating some logic, I wonder if this code should be refactored of having explicit functions for each of the possible cases. Right now it requires you to map the matrix of these booleans combinations
```
	candidate = findPossibleCandidateFromListWithRestrictions(intermediateSource, prevPrimary, preferredCandidates, true, true)
	candidate = findPossibleCandidateFromListWithRestrictions(intermediateSource, prevPrimary, preferredCandidates, false, true)
	candidate = findPossibleCandidateFromListWithRestrictions(intermediateSource, prevPrimary, preferredCandidates, true, false)
```

I wonder if it will be better to have something like:

```
	// check whether the one we promoted is in the same cell and belongs to the preferred candidates list
	candidate = findCandidateSameCellAndPreferred(intermediateSource, prevPrimary, preferredCandidates)
        // Instead of: candidate = findPossibleCandidateFromListWithRestrictions(intermediateSource, prevPrimary, preferredCandidates, true, true)
```

What do other folks think about this?

Reply:
I have added comments to all boolean arguments for now.
Orchestrator has the function unrolled into about 8 for loops. I thought that this way would be more concise, clean and readable but if it is not serving that purpose then we can change it to be a bunch of repeated for loops like orchestrator or as separate 4 functions which correspond to the 4 possible variations [(true,true), (true,false), (false,false), (true,true)]. I do not have any strong opinions about it

Reply:
I have made a change to this code and now we have two functions, `findCandidateAnyCell` and `findCandidateSameCell`. I hope it is more readable now

Reply:
This makes sense!